### PR TITLE
use shared informers for BootstrapSigner controller

### PR DESCRIPTION
**Pull request by wackxu**
**What this PR does / why we need it**:

fix TODO: Switch to shared informers

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by deads2k** [comment]
The idea is sound.  I think this controller was for `kubeadm`, I'll let @luxas do the detailed review

/approve no-issue
/assign luxas
/unassign

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (luxas) [inline pkg/controller/bootstrap/bootstrapsigner.go]: shouldn't something be logged here?

**Review by luxas — APPROVED** [review]
Thanks! I didn't see any critical things that needed changing, but I didn't write the controller so I might have missed something

/assign @jbeda 
that wrote this :)

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: @luxas Actually only if  `stopCh` is closed it will return, that is all controller is closed. I follow other controller forms and do not log here.If actually need, I can add this.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (zjj2wry) [inline pkg/controller/bootstrap/bootstrapsigner.go]: we can use func like `cache.DeletionHandlingMetaNamespaceKeyFunc` get configmap key?

Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: @zjj2wry yeah, we can. But `e.configMapNamespace` and `e.configMapName` here is always not changed, so we can simply  the logic

Comment (zjj2wry) [reply pkg/controller/bootstrap/bootstrapsigner.go]: got it, this lgtm

**Comment by wackxu** [comment]
kindly ping @jbeda for approval. Also with #55701

**Comment by jbeda** [comment]
Thanks for doing this!  I'm rusty at this part of the code base but it looks good.

/lgtm

**Comment by luxas** [comment]
> I'm rusty at this part of the code

/hold

Let's get @ncdc's approval as well.
Also, is there any huge advantage of having this in v1.9 vs in v1.10 given that this feature will stay unchanged in the beta state in v1.9?

/assign @ncdc

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Logging is done by `controller.WaitForCacheSync`.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: Do you need to match on t.Namespace too?

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: Do you need to match on t.Namespace?

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: I think you should be listing only in the appropriate namespace, right? This lists across all namespaces.

**Review by ncdc — COMMENTED** [review]
Just did a quick pass.

**Comment by wackxu** [comment]
@ncdc Thanks for your review. Done, PTAL

**Comment by wackxu** [comment]
@ncdc Can you also take a look  at  #55701

**Comment by wackxu** [comment]
@ncdc @luxas PTAL, I think it is time to merge it.

**Comment by errordeveloper** [comment]
LGTM, if @ncdc approves it.

**Comment by ncdc** [comment]
Will look when I’m at my desk in a couple of hours

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: Please set these as part of the initial declaration of `e`. We need to avoid the situation where the event handler fires before the lister is set.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: Please set these as part of the initial declaration of `e`. We need to avoid the situation where the event handler fires before the lister is set.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: My personal preference is to group the lister & synced vars together, i.e.
```go
secretLister    corelisters.SecretLister
secretSynced    cache.InformerSynced
```

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: I would keep `configMapKey` and set it as part of `NewBootstrapSigner`. That way you can remove `getConfigMapKey` and just reference `configMapKey` directly.

You'll also need to keep `configMapName` and `configMapNamespace` as you've added them, so you can do the filtering.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: Do you need to add checks to make sure the required fields in `options` have values?

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: I think the previous code would return error=nil and exists=false if the ConfigMap didn't exist. In that case, it wouldn't call `utilruntime.HandleError` because err is nil.

The new code will return a NotFound error if it doesn't exist. Do you want to check for that and return nil without calling HandleError in that case?

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: Is this ok? It means that `getTokens` gets no secrets, which means that `signConfigMap` gets no tokens. Is that alright, or does the error need to be propagated up?

**Review by ncdc — CHANGES_REQUESTED** [review]

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Sounds good, Done

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: I think it is not necessary, It is just called by `startBootstrapSignerController` and the parameter is the default parameter `bootstrap.DefaultBootstrapSignerOptions()`, we just keep as the original

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Great， updated

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Updated

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Done

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Done

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Yes, we need just as `getConfigMap` when there has an error, It is also keep as the original logic

**Comment by wackxu** [comment]
All comments are addressed or clarified, @ncdc  PTAL

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [inline pkg/controller/bootstrap/bootstrapsigner.go]: I would probably do this as
```go
if err != nil {
  if !apierrors.IsNotFound(err) {
    utilruntime.HandleError(err)
  }
  return nil
}

return configMap
```

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (ncdc) [reply pkg/controller/bootstrap/bootstrapsigner.go]: There is a difference, though, between the old code and this PR. Before, it was impossible for `e.secrets.List()` to return an error. Now, when using `secretLister`, you may get an error. Just wanted to point that out.

File: pkg/controller/bootstrap/bootstrapsigner.go
Comment (wackxu) [reply pkg/controller/bootstrap/bootstrapsigner.go]: Make sense, Updated

**Review by ncdc — APPROVED** [review]

**Pull request merged**
